internal/mode/static/state/graph/policies.go

```diff
@@ -0,0 +1,310 @@
+		case kinds.HTTPRoute, kinds.GRPCRoute:
+			route, exists := g.Routes[routeKeyForKind(ref.Kind, ref.Nsname)]
+			if !exists {
+				continue
+			}
+
+			attachPolicyToRoute(policy, route, ctlrName)
+		}
+	}
+}
+
+func attachPolicyToRoute(policy *Policy, route *L7Route, ctlrName string) {
+	kind := v1.Kind(kinds.HTTPRoute)
+	if route.RouteType == RouteTypeGRPC {
+		kind = kinds.GRPCRoute
+	}
+
+	routeNsName := types.NamespacedName{Namespace: route.Source.GetNamespace(), Name: route.Source.GetName()}
+
+	ancestor := &PolicyAncestor{
+		Ancestor: createParentReference(v1.GroupName, kind, routeNsName),
+	}
+
+	curAncestorStatus := policy.Source.GetPolicyStatus().Ancestors
+	if ancestorsFull(curAncestorStatus, ancestor.Ancestor, ctlrName) {
+		return
+	}
+
+	policy.Ancestor = ancestor
+
+	if !route.Valid || !route.Attachable || len(route.ParentRefs) == 0 {
+		policy.Ancestor.Conditions = []conditions.Condition{staticConds.NewPolicyTargetNotFound("TargetRef is invalid")}
+
+		return
+	}
+
+	route.Policies = append(route.Policies, policy)
+}
+
+func attachPolicyToGateway(
+	policy *Policy,
+	gw *Gateway,
+	ignoredGateways map[types.NamespacedName]*v1.Gateway,
+	ctlrName string,
+) {
+	ref := policy.TargetRef
+
+	_, ignored := ignoredGateways[ref.Nsname]
+
+	if !ignored && ref.Nsname != client.ObjectKeyFromObject(gw.Source) {
+		return
+	}
+
+	ancestor := &PolicyAncestor{
+		Ancestor: createParentReference(v1.GroupName, kinds.Gateway, ref.Nsname),
+	}
+
+	curAncestorStatus := policy.Source.GetPolicyStatus().Ancestors
+	if ancestorsFull(curAncestorStatus, ancestor.Ancestor, ctlrName) {
+		return
+	}
+
+	policy.Ancestor = ancestor
+
+	if ignored {
+		policy.Ancestor.Conditions = []conditions.Condition{staticConds.NewPolicyTargetNotFound("TargetRef is ignored")}
+		return
+	}
+
+	if !gw.Valid {
+		policy.Ancestor.Conditions = []conditions.Condition{staticConds.NewPolicyTargetNotFound("TargetRef is invalid")}
+		return
+	}
+
+	gw.Policies = append(gw.Policies, policy)
+}
+
+func processPolicies(
+	policies map[PolicyKey]policies.Policy,
+	validator validation.PolicyValidator,
+	gateways processedGateways,
+	routes map[RouteKey]*L7Route,
+) map[PolicyKey]*Policy {
+	if len(policies) == 0 || gateways.Winner == nil {
+		return nil
+	}
+
+	processedPolicies := make(map[PolicyKey]*Policy)
+
+	for key, policy := range policies {
+		ref := policy.GetTargetRef()
+		refNsName := targetRefNsName(ref, policy.GetNamespace())
+
+		refGroupKind := fmt.Sprintf("%s/%s", ref.Group, ref.Kind)
+
+		switch refGroupKind {
+		case gatewayGroupKind:
+			if !gatewayExists(refNsName, gateways.Winner, gateways.Ignored) {
+				continue
+			}
+		case hrGroupKind, grpcGroupKind:
+			if _, exists := routes[routeKeyForKind(ref.Kind, refNsName)]; !exists {
+				continue
+			}
+		default:
+			continue
+		}
+
+		var conds []conditions.Condition
+
+		if err := validator.Validate(policy); err != nil {
+			conds = append(conds, staticConds.NewPolicyInvalid(err.Error()))
```

Comment:
1. I think this is fine. The duplicate code is minimal, and returning conditions makes sense in the long run since policies will likely have different conditions. 
2. I agree that the specific validator's responsibility is to handle this special logic. I wonder if we can connect the observability policy validator to the nginx proxy in the graph so we don't have to make a k8s api call. What if we did something similar to enabling leader election in the status updater? We connect the two with a callback - either a setter (set allow/disallow telemetry on the validator) or a getter (get the nginx proxy from the graph).

Reply:
I wonder if using the graph could be an issue though, because this code is currently building the graph. It may be too flimsy to look up the state of the graph while we're building it, because in the future we may change the order of resources we populate in the graph.

Reply:
That's a good point.

 I don't think we can rely on the k8s api alone because even if the NginxProxy exists and is referenced by the GatewayClass, how do we know that the NginxProxy resource is valid?

Reply:
We could check the status of the GatewayClass to see. But I think there's an edge case where NginxProxy and ObservabilityPolicy are both in the same batch, so the status hasn't been updated yet for us to see.

Reply:
We could also remove the dependency between the two, and in our tracing/troubleshooting guide, say that if tracing isn't working, be sure to check the status of both resources.

Reply:
Yep, that is def an edge case.

I think we need to guarantee the ordering of graph processing. Policies have to be processed last because they rely on the status of the other resources.

Reply:
> We could also remove the dependency between the two, and in our tracing/troubleshooting guide, say that if tracing isn't working, be sure to check the status of both resources.

I don't hate that.

Reply:
Nevermind, just tested that the directives used in the policy only work if the directives in the NginxProxy are configured (otherwise the nginx module errors out).